Implemented Fling (1/2)

LOuroboros opened the pull request:
## Description
Closes #50

I worked on this all from November 2020 to January 2021. With the back then pending revamp of the item_expansion I put it on hold, but now that the revamp happened and it's merged, I think it's time to submit what I have.
The commit history does not reflect this because I reimplemented it on a clean copy of the BE, and then I changed some things around, filling in some blanks, updating things such as the values of `waitmessage` calls and stuff like that, and then I force pushed it all.
A companion PR for the item_expansion branch is coming up in a moment.

## **Discord contact info**
Lunos#4026

AsparagusEduardo commented on src/battle_script_commands.c:
```diff
@@ -9272,6 +9285,89 @@ static void Cmd_various(void)
+        else if (gLastUsedItem == ITEM_NONE)
+    #else
+        if (gLastUsedItem == ITEM_NONE)
+    #endif
```
IMO, `ItemId_GetFlingPower` and `flingPower`should be added by both BE and IE. That way, we allow people who only use BE to also be able to use Fling (they would only need to add the data to `gItems`.

LOuroboros replied:
I don't agree with this because the functions are item specific as I mentioned on Discord, but also now that I'm checking, the changes are bound to cause at least 2 merge conflicts *(`include/item.h` and `src/item.c`)* due to the placement of `ItemId_GetFlingPower` and the existence of `IsPinchBerryItemEffect` in the battle_engine, which does not exist in the item_expansion.

Because of that, I rather wait for a second opinion :eyes:

AsparagusEduardo commented on src/battle_script_commands.c:
```diff
@@ -9272,6 +9285,89 @@ static void Cmd_various(void)
             gBattlescriptCurrInstr += 7;
         return;
     }
+    case VARIOUS_JUMP_IF_FLING_FAILS:
```
There's a placeholder function called `CanFling` which I feel it should be used instead of some of these checks, though it should be updated to remove checks already handled by
`if (!ItemId_GetFlingPower(gBattleMons[gActiveBattler].item))`.

It should be noted that there's already the check for Unnerve in there, so there's no need for the workaround in `attackcanceler` but it has to be fixed so it uses `IsUnnerveAbilityOnOpposingSide`.

AsparagusEduardo replied:
Better to use `UnnerveOn`, as that already has the check for the berry pocket.

LOuroboros replied:
I'll look into it :eyes:

LOuroboros replied:
> Better to use `UnnerveOn`, as that already has the check for the berry pocket.

Alright, so to clarify, by this you mean you want me to replace the `IsAbilityOnSide(battlerId, ABILITY_UNNERVE)` check inside of `CanFling` with `UnnerveOn(battlerId, item)` instead, correct? :eyes:

LOuroboros replied:
> but it has to be fixed so it uses `IsUnnerveAbilityOnOpposingSide`.

I don't understand this part. It's an attacker with Unnerve who's not supposed to be able to use Fling if they're holding a berry, not an opponent 😅

AsparagusEduardo replied:
Source? Bulbapedia only says that Unnerve prevents flinging berries, and in every other instancUnnerve only affects the opposing side, so it makes sense that this would also affect the opposing side.

LOuroboros replied:
> Source? Bulbapedia only says that Unnerve prevents flinging berries

No, it doesn't say "just" that 😅 

> Unnerve prevents a Pokémon holding a Berry from using Fling.

[That's specifically what it says](https://bulbapedia.bulbagarden.net/wiki/Fling_(move)).

AsparagusEduardo replied:
But it doesn't specify if the mon with Unnerve is the same one as the one holding the berry, it just says *a* Pokémon, not *which* one (itself, partner or opponents).
The phrasing is kinda off 🤔

LOuroboros replied:
> But it doesn't specify if the mon with Unnerve is the same one as the one holding the berry, it just says _a_ Pokémon, not _which_ one (itself, partner or opponents). The phrasing is kinda off 🤔

It says quite clearly that a Pokémon with Unnerve holding a berry can't use Fling, which is a move. I don't know what's hard to understand about that, to be quite honest with you.

LOuroboros commented:
After tests on ORAS and USUM, and also some chatting on Discord, I decided to push what I had.

AsparagusEduardo commented:
`CanFling` still needs to have the `ItemId_GetFlingPower` check. With it, there's not gonna be a need for some checks.
```diff
    if (item == ITEM_NONE
      #if B_KLUTZ_FLING_INTERACTION >= GEN_5
      || GetBattlerAbility(battlerId) == ABILITY_KLUTZ
      #endif
      || gFieldStatuses & STATUS_FIELD_MAGIC_ROOM
      || gDisableStructs[battlerId].embargoTimer != 0
+     || ItemId_GetFlingPower(item) != 0
+     || !CanBattlerGetOrLoseItem(battlerId, item))
-     || !CanBattlerGetOrLoseItem(battlerId, item)
-     //|| itemEffect == HOLD_EFFECT_PRIMAL_ORB
-     || itemEffect == HOLD_EFFECT_GEMS
-     #ifdef ITEM_ABILITY_CAPSULE
-     || item == ITEM_ABILITY_CAPSULE
-     #endif
-     || GetPocketByItemId(item) == POCKET_POKE_BALLS)
        return FALSE;
```

LOuroboros commented:
> `CanFling` still needs to have the `ItemId_GetFlingPower` check. With it, there's not gonna be a need for some checks.
> 
> ```diff
>     if (item == ITEM_NONE
>       #if B_KLUTZ_FLING_INTERACTION >= GEN_5
>       || GetBattlerAbility(battlerId) == ABILITY_KLUTZ
>       #endif
>       || gFieldStatuses & STATUS_FIELD_MAGIC_ROOM
>       || gDisableStructs[battlerId].embargoTimer != 0
> +     || ItemId_GetFlingPower(item) != 0
> +     || !CanBattlerGetOrLoseItem(battlerId, item))
> -     || !CanBattlerGetOrLoseItem(battlerId, item)
> -     //|| itemEffect == HOLD_EFFECT_PRIMAL_ORB
> -     || itemEffect == HOLD_EFFECT_GEMS
> -     #ifdef ITEM_ABILITY_CAPSULE
> -     || item == ITEM_ABILITY_CAPSULE
> -     #endif
> -     || GetPocketByItemId(item) == POCKET_POKE_BALLS)
>         return FALSE;
> ```

Done :eyes:

LOuroboros commented:
I'm sick of waiting so I'll just push what I have. I moved `IsPinchBerryItemEffect` to a more suitable file, the only one where it's used, `src/battle_ai_main.c`. This was done in order to deal with the conflicts at `include/item.h` and `src/item.c` and solved the conflicts.

I also introduced the `flingPower` item field to the BE, even though I still don't like the idea personally.

AsparagusEduardo commented on src/battle_util.c:
```diff
@@ -7964,7 +7964,9 @@ static u16 CalcMoveBasePower(u16 move, u8 battlerAtk, u8 battlerDef)
         break;
     case EFFECT_FLING:
-        // todo: program Fling + Unburden interaction
+    #ifdef ITEM_EXPANSION
```
Now that `ItemId_GetFlingPower` is part of the PR, there's no need for the item expansion check

AsparagusEduardo commented on src/battle_util.c:
```diff
@@ -9637,17 +9639,15 @@ bool32 CanFling(u8 battlerId)
-      #ifdef ITEM_ABILITY_CAPSULE
-      || item == ITEM_ABILITY_CAPSULE
+      #ifdef ITEM_EXPANSION
+      || ItemId_GetFlingPower(item) != 0
```
Same as above

LOuroboros commented:
Ready for re-re-review :eyes:
